server/s3_object_data_copier.h

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2020 Seagate Technology LLC and/or its Affiliates
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+#pragma once
+
+#include <cstddef>
+#include <functional>
+#include <memory>
+#include <string>
+
+#include <gtest/gtest_prod.h>
+
+#include "lib/types.h"  // struct m0_uint128
+#include "s3_buffer_sequence.h"
+
+class RequestObject;
+class S3MotrReader;
+class S3MotrReaderFactory;
+class S3MotrWiter;
+
+class S3ObjectDataCopier {
+
+  std::string request_id;
+  std::string s3_error;
+
+  std::function<bool(void)> check_shutdown_and_rollback;
+  std::function<void(void)> on_success;
+  std::function<void(void)> on_failure;
+
+  std::shared_ptr<RequestObject> request_object;
+  std::shared_ptr<S3MotrWiter> motr_writer;
+  std::shared_ptr<S3MotrReaderFactory> motr_reader_factory;
+  std::shared_ptr<MotrAPI> motr_api;
+  std::shared_ptr<S3MotrReader> motr_reader;
+
+  S3BufferSequence data_blocks_read;     // Source's data has been read but not
+                                         // taken for writing.
+  S3BufferSequence data_blocks_writing;  // Source's data currently beeing
+                                         // written
+
+  // All POD variables should be (re)initialized in This::copy()
+  size_t bytes_left_to_read;
+  size_t motr_unit_size;
+  bool copy_failed;
+  bool read_in_progress;
+  bool write_in_progress;
+
+  void cleanup_blocks_written();
+  void read_data_block();
+  void read_data_block_success();
+  void read_data_block_failed();
+  void write_data_block();
+  void write_data_block_success();
+  void write_data_block_failed();
+  void set_s3_error(std::string);
+
+ public:
+  S3ObjectDataCopier(std::shared_ptr<RequestObject> request_object,
```

Comment:
S3ObjectDataCopier can abstract details of motr APIs. 
S3ObjectDataCopier(std::shared_ptr<RequestObject> request_object, string object_store_interface = "motr");
Internally it can create required motr structures like motr_writer and reader etc

Reply:
Discussed with Evgeniy, we can add a constructor later